# 01 Extraction: Data Audit

**Purpose:** You are a detective. You are not touching the crime scene. You are just documenting it.

This notebook loads the raw dataset, inspects its structure, and documents baseline conditions. No cleaning or modification happens here.

## 1. Import Libraries
Importing pandas, numpy, matplotlib, and seaborn for data loading and initial inspection. Nothing else is needed here.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from pathlib import Path

## 2. Load the CSV
Loading from `data/raw/`. Confirming the shape matches expectations (e.g., 30,000 × 25).

In [2]:
raw_path = Path('../data/raw/UCI_Credit_Card_raw.csv')
df = pd.read_csv(raw_path)
print(f'Shape of the dataset: {df.shape}')

Shape of the dataset: (30000, 25)


## 3. Print Column Names and Data Types
Checking if any numerical column loaded as an object (string). That would break calculations later.

In [3]:
print(df.dtypes)

ID                              int64
LIMIT_BAL                     float64
SEX                             int64
EDUCATION                     float64
MARRIAGE                      float64
AGE                             int64
PAY_0                           int64
PAY_2                           int64
PAY_3                           int64
PAY_4                           int64
PAY_5                           int64
PAY_6                           int64
BILL_AMT1                     float64
BILL_AMT2                     float64
BILL_AMT3                     float64
BILL_AMT4                     float64
BILL_AMT5                     float64
BILL_AMT6                     float64
PAY_AMT1                      float64
PAY_AMT2                      float64
PAY_AMT3                      float64
PAY_AMT4                      float64
PAY_AMT5                      float64
PAY_AMT6                      float64
default.payment.next.month      int64
dtype: object


## 4. Head and Tail
Checking the first 5 rows and last 5 rows to ensure the data looks sensible end to end.

In [4]:
display(df.head())
display(df.tail())

,ID,LIMIT_BAL,SEX,EDUCATION,MARRIAGE,AGE,PAY_0,PAY_2,PAY_3,PAY_4,...,BILL_AMT4,BILL_AMT5,BILL_AMT6,PAY_AMT1,PAY_AMT2,PAY_AMT3,PAY_AMT4,PAY_AMT5,PAY_AMT6,default.payment.next.month
0,1,20000.0,2,2.0,1.0,24,2,2,-1,-1,...,0.0,0.0,0.0,0.0,689.0,0.0,0.0,0.0,0.0,1
1,2,120000.0,2,2.0,2.0,26,-1,2,0,0,...,3272.0,3455.0,3261.0,0.0,1000.0,1000.0,1000.0,0.0,2000.0,1
2,3,90000.0,2,2.0,2.0,34,0,0,0,0,...,14331.0,14948.0,15549.0,1518.0,1500.0,1000.0,1000.0,1000.0,5000.0,0
3,4,50000.0,2,2.0,1.0,37,0,0,0,0,...,28314.0,28959.0,29547.0,2000.0,2019.0,1200.0,1100.0,1069.0,1000.0,0
4,5,50000.0,1,2.0,1.0,57,-1,0,-1,0,...,20940.0,19146.0,19131.0,2000.0,36681.0,10000.0,9000.0,689.0,679.0,0


,ID,LIMIT_BAL,SEX,EDUCATION,MARRIAGE,AGE,PAY_0,PAY_2,PAY_3,PAY_4,...,BILL_AMT4,BILL_AMT5,BILL_AMT6,PAY_AMT1,PAY_AMT2,PAY_AMT3,PAY_AMT4,PAY_AMT5,PAY_AMT6,default.payment.next.month
29995,29996,220000.0,1,3.0,1.0,39,0,0,0,0,...,88004.0,31237.0,15980.0,8500.0,20000.0,5003.0,3047.0,5000.0,1000.0,0
29996,29997,150000.0,1,3.0,2.0,43,-1,-1,-1,-1,...,8979.0,5190.0,0.0,1837.0,3526.0,8998.0,129.0,0.0,0.0,0
29997,29998,30000.0,1,2.0,2.0,37,4,3,2,-1,...,20878.0,20582.0,19357.0,0.0,0.0,22000.0,4200.0,2000.0,3100.0,1
29998,29999,80000.0,1,3.0,1.0,41,1,-1,0,0,...,52774.0,11855.0,48944.0,85900.0,3409.0,1178.0,1926.0,52964.0,1804.0,1
29999,30000,50000.0,1,2.0,1.0,46,0,0,0,0,...,36535.0,32428.0,15313.0,2078.0,1800.0,1430.0,1000.0,1000.0,1000.0,1


## 5. Statistical Summary
Running `describe()` to check statistical summaries of all columns. 
Looking for: Are there negative LIMIT_BAL values? What is the AGE range? What are the min/max values of BILL_AMT?

In [5]:
display(df.describe())

,ID,LIMIT_BAL,SEX,EDUCATION,MARRIAGE,AGE,PAY_0,PAY_2,PAY_3,PAY_4,...,BILL_AMT4,BILL_AMT5,BILL_AMT6,PAY_AMT1,PAY_AMT2,PAY_AMT3,PAY_AMT4,PAY_AMT5,PAY_AMT6,default.payment.next.month
count,30000.000000,30000.000000,30000.000000,29655.000000,29946.000000,30000.000000,30000.000000,30000.000000,30000.000000,30000.000000,...,30000.000000,30000.000000,30000.000000,30000.000000,3.000000e+04,30000.00000,30000.000000,30000.000000,30000.000000,30000.000000
mean,15000.500000,167484.322667,1.603733,1.817164,1.554665,35.485500,0.075267,-0.007700,-0.030033,-0.075733,...,43262.948967,40311.400967,38871.760400,5663.580500,5.921163e+03,5225.68150,4826.076867,4799.387633,5215.502567,0.221200
std,8660.398374,129747.661567,0.489129,0.711283,0.518259,9.217904,0.990775,1.035798,1.025036,0.987436,...,64332.856134,60797.155770,59554.107537,16563.280354,2.304087e+04,17606.96147,15666.159744,15278.305679,17777.465775,0.415062
min,1.000000,10000.000000,1.000000,1.000000,1.000000,21.000000,-1.000000,-1.000000,-1.000000,-1.000000,...,-170000.000000,-81334.000000,-339603.000000,0.000000,0.000000e+00,0.00000,0.000000,0.000000,0.000000,0.000000
25%,7500.750000,50000.000000,1.000000,1.000000,1.000000,28.000000,-1.000000,-1.000000,-1.000000,-1.000000,...,2326.750000,1763.000000,1256.000000,1000.000000,8.330000e+02,390.00000,296.000000,252.500000,117.750000,0.000000
50%,15000.500000,140000.000000,2.000000,2.000000,2.000000,34.000000,0.000000,0.000000,0.000000,0.000000,...,19052.000000,18104.500000,17071.000000,2100.000000,2.009000e+03,1800.00000,1500.000000,1500.000000,1500.000000,0.000000
75%,22500.250000,240000.000000,2.000000,2.000000,2.000000,41.000000,0.000000,0.000000,0.000000,0.000000,...,54506.000000,50190.500000,49198.250000,5006.000000,5.000000e+03,4505.00000,4013.250000,4031.500000,4000.000000,0.000000
max,30000.000000,1000000.000000,2.000000,4.000000,3.000000,79.000000,8.000000,8.000000,8.000000,8.000000,...,891586.000000,927171.000000,961664.000000,873552.000000,1.684259e+06,896040.00000,621000.000000,426529.000000,528666.000000,1.000000


## 6. Check for Missing Values
Checking for null values. This documents our baseline cleanliness.

In [6]:
print(df.isnull().sum())

ID                              0
LIMIT_BAL                       0
SEX                             0
EDUCATION                     345
MARRIAGE                       54
AGE                             0
PAY_0                           0
PAY_2                           0
PAY_3                           0
PAY_4                           0
PAY_5                           0
PAY_6                           0
BILL_AMT1                       0
BILL_AMT2                       0
BILL_AMT3                       0
BILL_AMT4                       0
BILL_AMT5                       0
BILL_AMT6                       0
PAY_AMT1                        0
PAY_AMT2                        0
PAY_AMT3                        0
PAY_AMT4                        0
PAY_AMT5                        0
PAY_AMT6                        0
default.payment.next.month      0
dtype: int64


## 7. Check for Duplicate Rows
Checking for exact duplicate rows in the dataset. Expected: 0.

In [7]:
print(f'Duplicate rows: {df.duplicated().sum()}')

Duplicate rows: 0


## 8. Value Counts on Categorical Columns
Running value counts on specific categorical columns where dirty/undocumented values might live.
- **EDUCATION**
- **MARRIAGE**
- **SEX**
- **default.payment.next.month**

In [8]:
cols_to_check = ['EDUCATION', 'MARRIAGE', 'SEX', 'default.payment.next.month']
for col in cols_to_check:
    print(f'\n--- Value counts for {col} ---')
    print(df[col].value_counts(dropna=False))


--- Value counts for EDUCATION ---
EDUCATION
2.0    14030
1.0    10585
3.0     4917
NaN      345
4.0      123
Name: count, dtype: int64

--- Value counts for MARRIAGE ---
MARRIAGE
2.0    15964
1.0    13659
3.0      323
NaN       54
Name: count, dtype: int64

--- Value counts for SEX ---
SEX
2    18112
1    11888
Name: count, dtype: int64

--- Value counts for default.payment.next.month ---
default.payment.next.month
0    23364
1     6636
Name: count, dtype: int64


## 9. Class Imbalance Check
Calculating what percentage of rows are default=1 vs default=0.

In [ ]:
default_counts = df['default.payment.next.month'].value_counts(normalize=True) * 100
print('Class Distribution (Percentage):') 
print(default_counts)

Class Distribution (Percentage):
default.payment.next.month
0    77.88
1    22.12
Name: proportion, dtype: float64


> **Observation**: The dataset exhibits a class imbalance. Approximately 77.88% of the clients did not default (class `0`), while 22.12% defaulted (class `1`).

## 10. Final Summary: Data Description

### What We Found
- **Shape**: The dataset contains 30,000 rows and 25 columns.
- **Data Types**: All features correctly loaded as numerical (int64/float64). No strings object conversions were found that would break calculations.
- **Duplicates**: There are no duplicate rows.
- **Target Imbalance**: The target variable `default.payment.next.month` is imbalanced (~78% vs ~22%).

### Problems Identified (To be fixed in the Cleaning phase)
- **Missing Values**: We discovered missing values (NaN) in `EDUCATION` and `MARRIAGE` representing undocumented entries.
- **Undocumented Categories**: We may have observed unexpected categories such as `0`, `5`, or `6` in `EDUCATION` (if not mapped to NaN) or undocumented negative payments in `PAY_X`.

*Note: This raw data is untouched. All data cleaning, imputation, and category collapsing will occur in Notebook 02.*